### PR TITLE
3CX check for updates

### PR DESCRIPTION
**Pull request by CPbN**
Hi,

This PR enhances 3CX plugin checking for internal software updates.

Thank you 👍

### CODE REVIEW

File: apps/voip/3cx/restapi/custom/api.pm [hunk 1]
```diff
@@ -194,6 +194,12 @@ sub request_api {
         warning_status => '', unknown_status => '', critical_status => '%{http_code} < 200 or %{http_code} >= 300'
     );
 
+    # Some content may be strangely returned, for example : "[{\"Category\":\"provider\",\"Count\":1}]"
+    if ($content =~ /^"(\[.*\])"$/) {
+        $content = $1;
+        $content =~ s/\\"/"/g;
+    }
```
Comment (garnier-quentin) [inline apps/voip/3cx/restapi/custom/api.pm]: I can't accept that. It can break json

File: apps/voip/3cx/restapi/mode/system.pm [hunk 2]
```diff
@@ -96,8 +96,8 @@ sub new {
     $self->{version} = '1.0';
     $options{options}->add_options(arguments => {
         "unknown-status:s"  => { name => 'unknown_status', default => '' },
-        "warning-status:s"  => { name => 'warning_status', default => '' },
-        "critical-status:s" => { name => 'critical_status', default => '%{health} =~ /false/' },
+        "warning-status:s"  => { name => 'warning_status', default => '%{display} =~ /HasUpdatesAvailable/ && %{health} =~ /true/' },
+        "critical-status:s" => { name => 'critical_status', default => '%{display} !~ /HasUpdatesAvailable/ && %{health} =~ /false/' },
```
Comment (garnier-quentin) [inline apps/voip/3cx/restapi/mode/system.pm]: Not by default. BUt you can set the threshold in your command line.

Comment (CPbN) [reply apps/voip/3cx/restapi/mode/system.pm]: The previous `critical-status` would then trigger if no update is available, as `HasUpdatesAvailable` would then be false.

Would I then have to rename `HasUpdatesAvailable` to `SystemIsUptodate` ?
An available update would then trigger a critical alert.

File: apps/voip/3cx/restapi/custom/api.pm (hunk 1 above)
Comment (CPbN) [reply apps/voip/3cx/restapi/custom/api.pm]: The added API call returns `"[{\"Category\":\"provider\",\"Count\":1}]"`.
This is not JSON compliant, this is why we have to remove leading and trailing `"`, and remove escaping `\` characters before `"`, but of course only if `$content` matches this strange content format.

File: apps/voip/3cx/restapi/mode/system.pm (hunk 2 above)
Comment (garnier-quentin) [reply apps/voip/3cx/restapi/mode/system.pm]: Yes that better. like that, we use the same for 'true' and 'false'.

File: apps/voip/3cx/restapi/custom/api.pm (hunk 1 above)
Comment (garnier-quentin) [reply apps/voip/3cx/restapi/custom/api.pm]: The JSON is not compliant ? So that's a bug of 3cx ?

Comment (CPbN) [reply apps/voip/3cx/restapi/custom/api.pm]: I think I will not manage to get 3CX to correct this, as it is not documented nor supported :-|
In addition, this is not really a bug, as the JSON string is simply surrounded by `"`, which could in some kind be "expected". We simply have to remove the surrounding...

File: apps/voip/3cx/restapi/mode/system.pm (hunk 2 above)
Comment (CPbN) [reply apps/voip/3cx/restapi/mode/system.pm]: OK, let me improve this then :) :+1:

Comment (CPbN) [reply apps/voip/3cx/restapi/mode/system.pm]: Should be OK now Quentin :)

File: apps/voip/3cx/restapi/custom/api.pm (hunk 1 above)
Comment (garnier-quentin) [reply apps/voip/3cx/restapi/custom/api.pm]: Could you provide me the raw string result of that API call ?

Comment (CPbN) [reply apps/voip/3cx/restapi/custom/api.pm]: Here it is :
`"[{\"Category\":\"provider\",\"Count\":1}]"`

Comment (CPbN) [reply apps/voip/3cx/restapi/custom/api.pm]: If no update is available, I get :
`"[]"`

Comment (garnier-quentin) [reply apps/voip/3cx/restapi/custom/api.pm]: The string seems ok. You should can decode with json without problem (no change is needed)

Comment (garnier-quentin) [reply apps/voip/3cx/restapi/custom/api.pm]: You do the print with Data::Dumper ?

Comment (CPbN) [reply apps/voip/3cx/restapi/custom/api.pm]: Without this trick the exception is triggered :
`Cannot decode json response: JSON text must be an object or array (but found number, string, true, false or null, use allow_nonref to allow this)`

Comment (CPbN) [reply apps/voip/3cx/restapi/custom/api.pm]: I print response content using `--debug`.
This is also what curl returns.

Comment (CPbN) [reply apps/voip/3cx/restapi/custom/api.pm]: With Data::Dumper :
`$VAR1 = '"[{\\"Category\\":\\"provider\\",\\"Count\\":1}]"';`

**Comment by CPbN** [comment]
Many thx Quentin 👍

**Comment by garnier-quentin** [comment]
I have changed your code. Could you test it ?

**Comment by CPbN** [comment]
Works nicely, nice trick !
Thx 👍

**Pull request merged**